### ISSUE TITLE
[ML] Adding an endpoint for plotting source data for an anomaly detector

### ISSUE THREAD

**Issue by droberts195** [elastic/elasticsearch#66131]
When investigating an anomaly it is useful to look back at the source data for the time series that the anomaly was detected in around the time it was detected.

The ML UI contains logic to build an appropriate search to plot this source data on drilldown charts when needed.

However, as the number of ways to configure the datafeed search increases, so this UI logic becomes more complex.

There would be benefits in adding a backend endpoint that would get the numbers to plot on the drilldown chart:

- Simplified UI code
- It would move the job of supporting new ES features, e.g. runtime fields, to the backend team
- In the long run we could more easily add support for functions that cannot currently be plotted, such as `info_content`
- OEMs who integrate with ML in Elasticsearch but develop their own UI could use the endpoint
- The numbers used in drilldown functionality could be tested in isolation of the UI

The endpoint would take a request containing the following information:

1. Job ID
2. Detector index
3. Time range
4. Bucket span for drilldown points - this would be a multiple of the job's bucket span or possibly we could allow it to be a fraction
5. Split field values (by, over and/or partition as appropriate for the specified detector index)

The job ID would be used to look up the datafeed ID.  Alternatively the request could take the datafeed ID and that be used to look up the job ID.  (In the long run it would be nice if these IDs were the same.)

The response would contain the values to plot on the chart.  We would need to decide if this should be in the same format that an aggregation would return, or further simplified from that.  It depends on what would make life easiest for the ML UI.

@peteharverson please add any extra requirements to enable this endpoint to be used in the UI.

**Comment by benwtrent** [comment]
> Bucket span for drilldown points - this would be a multiple of the job's bucket span or possibly we could allow it to be a fraction

I don't think this is simple (especially no way on a fraction if that fraction is not a multiple of the datafeed histogram interval).

Consider the following situation:

- Data feed histogram interval is `5m`
- Aggregated value is a `max` value
- the detector is an `mean` on that aggregated `max` value
- The job bucket size is `15m`

Either:

- ES does some second level analytics on the results of the aggregated output
- Or the consumer does (in this case Kibana)

**Comment by droberts195** [comment]
> I don't think this is simple (especially if the fraction is not a multiple of the datafeed histogram interval).

@peteharverson what are the possible bucket spans used in UI drilldown charts at present?  We don't need to support anything that would never be requested.  We can limit the initial functionality of the new API to what the UI might do itself today.

> * Or the consumer does (in this case Kibana)

The whole point of this API is to reduce the UI to mindlessly plotting the points returned by the endpoint.  If there needs to be a second layer of processing in the UI then we shouldn't create this endpoint at all as it would result in higher total complexity across the stack.

**Comment by peteharverson** [comment]
The charts in the Anomaly Explorer will always plot at bucket span interval.

For the Single Metric Viewer, it will always be a multiple of the bucket span. We don't currently allow aggregation intervals which are fractions of the bucket span. The aggregation interval depends on the time range set in the view, and the width of the brush, and can potentially be any multiple (i.e. we _don't_ for example use fixed multiples of 2, 4, 8...), for example here is a job with a `17m` bucket span and the aggregation interval is 7 times the bucket span:

![image](https://user-images.githubusercontent.com/7405507/101767949-dd015400-3adc-11eb-895d-eef5d976f2b3.png)

One extra question comes to mind; do we want the endpoint to return model plot data, including bounds, if model plot has been enabled for the job?

**Comment by walterra** [comment]
> The response would contain the values to plot on the chart. We would need to decide if this should be in the same format that an aggregation would return, or further simplified from that. It depends on what would make life easiest for the ML UI.

Since it's a use case specific API IMHO it could be a simplified format. E.g. for some charts we create deeply nested aggs (Anomaly Explorer population charts, rare charts come to mind), but then we transform that into something like this:

https://github.com/elastic/elastic-charts/blob/master/src/utils/data_samples/test_anomaly_swim_lane.ts

```js
[
  {
    laneLabel: 'i-71a7f77b',
    time: 1572847200,
    value: 1.393504,
  },
  {
    laneLabel: 'i-71a7f77b',
    time: 1572892200,
    value: 98.22463906512188,
  },
...
]
```

**Comment by benwtrent** [comment]
> The whole point of this API is to reduce the UI to mindlessly plotting the points returned by the endpoint. If there needs to be a second layer of processing in the UI then we shouldn't create this endpoint at all as it would result in higher total complexity across the stack.

@droberts195 I am saying, either ES or Kibana will need to do a second layer of processing. It is not possible to handle the scenarios with aggs alone. 

ES already has the datafeed extraction code. This endpoint would effectively be:

- Scroll through the data Datafeed extraction (this would have extra query filtering for timewindow and possibly partition/by labels)
- Do second order analytics on the results based on the detector function

I think if we are to add this API to ES (since the work will be complex to handle all the detector function statistics). It would be nice if it was usable by more things than a single module (anomaly time series viewer). 

Can we call arbitrary APIs in drill downs?

**Comment by dimitris-athanasiou** [comment]
I think some aspect of this ties in with switching datafeed's to always work with aggregations.

An API that returns the data for plotting would definitely need to use aggregations. Thus it implies we can generate the right aggregation for each detector. If we have that, it's only another step to wrap this in composite aggs in order to make datafeeds work with aggs. With the exception of categorization where we need the raw messages.

**Comment by benwtrent** [comment]
> Thus it implies we can generate the right aggregation for each detector.

Given my previous example, I don't think this is possible. When a datafeed has aggs, the detector function is effectively a pipeline agg that runs on the results of previous aggregated values. Pipeline aggs are restricted in their use and which "layer" of the aggs they exist.

A second pass on the aggregated data post extraction will be required.
